src/store/messages/saga.ts

```diff
@@ -253,31 +253,50 @@ export function* createOptimisticPreview(channelId: string, optimisticMessage) {
 }
 
 export function* performSend(channelId, message, mentionedUserIds, parentMessage, optimisticId) {
-  const chatClient = yield call(chat.get);
+  let messageSentSuccessfully = false;
 
-  const messageCall = call(
-    [
-      chatClient,
-      chatClient.sendMessagesByChannelId,
-    ],
-    channelId,
-    message,
-    mentionedUserIds,
-    parentMessage,
-    null,
-    optimisticId
-  );
-  return yield sendMessage(messageCall, channelId, optimisticId);
+  try {
+    if (messageSentSuccessfully) {
+      return;
+    }
+
+    const chatClient = yield call(chat.get);
+    const messageCall = call(
+      [chatClient, chatClient.sendMessagesByChannelId],
+      channelId,
+      message,
+      mentionedUserIds,
+      parentMessage,
+      null,
+      optimisticId
+    );
+    const createdMessage = yield messageCall;
+    messageSentSuccessfully = true;
+
+    return createdMessage;
+  } catch (e) {
+    yield call(messageSendFailed, optimisticId);
+    return null;
+  }
 }
 
 export function* sendMessage(apiCall, channelId, optimisticId) {
+  let messageSentSuccessfully = false;
+
   try {
+    if (messageSentSuccessfully) {
+      return;
+    }
```

Comment:
i don't .. follow. will this ever execute? You've just declared `messageSentSuccessfully` to be `false`

Reply:
fixed now!